### PR TITLE
Refactor trusted and macro proof rules

### PR DESCRIPTION
**Pull request by ajreynol**
This PR conglomerates a majority of our trusted proof rules into single rule ProofRule::TRUST for clarity, and introduces a subcategory `TrustId` that captures the source of the trusted step.

This allows us to simplify the API ProofRule enumeration.

It also does some minor renaming of macro proof rules (`MACRO_` should be used for any rule that is not formally justifiable to external users but expected to be reconstructed internally).

### CODE REVIEW

File: src/proof/trust_id.h [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+     without a proof */
+  THEORY_PREPROCESS,
+  /** A lemma added during theory-preprocessing without a proof */
+  THEORY_PREPROCESS_LEMMA,
+  /** A expanding of definitions of the input formula made without a proof */
+  THEORY_EXPAND_DEF,
+  /** An axiom for an introduced witness term without a corresponding proof */
+  WITNESS_AXIOM,
+  /** A rewrite whose proof could not be reconstructed */
+  REWRITE_RCONS,
```
Comment (HanielB) [inline src/proof/trust_id.h:49]: I suggest we start moving towards the "elaboration" rather "reconstruction" terminology

File: src/smt/proof_post_processor.cpp [hunk 2]
```diff
@@ -959,7 +961,7 @@ Node ProofPostprocessCallback::expandMacros(ProofRule id,
                           << std::endl;
     return sumBounds;
   }
-  else if (id == ProofRule::STRING_INFERENCE)
+  else if (id == ProofRule::MACRO_STRING_INFERENCE)
```
Comment (HanielB) [inline src/smt/proof_post_processor.cpp:964]: If I understand correctly, the process here will add a theory inference trusted step when the elaboration fails, right?

**Review by HanielB — COMMENTED** [review]
The PR comment about macro rules not being formally justified to external users is too strong. Saying they *can* be internally elaborated suffices, I think. This is actually what is in the overall doc of the proof_rule module, which I think is fine.

File: src/smt/proof_post_processor.cpp (hunk 2 above)
Comment (ajreynol) [reply src/smt/proof_post_processor.cpp:964]: Yes.

**Review by HanielB — APPROVED** [review]

**Pull request merged**